### PR TITLE
feat: add support for HT1

### PR DESCRIPTION
**Pull request by ashleigh-hopkins**
Hey, I reverse engineered the calculation for the HT1 and thought it might be useful for some people with this older device. Tested with the single HT1 I own (I also have a HT.w).

I wrote some tests, but I'm not actually sure how to execute them - I'm pretty new to python 😅 

related issue https://github.com/home-assistant/core/issues/76638

### CODE REVIEW

File: src/sensorpush_ble/parser.py [hunk 1]
```diff
@@ -40,18 +41,56 @@
...
+
+
+def temperature_celsius_from_raw_temperature(num: int) -> float:
+    return round((-46.85) + (175.72 * (num / (pow(2.0, 14.0)))), 2)
+
+
 def decode_values(
     mfg_data: bytes, device_type_id: int
 ) -> dict[BaseSensorDescription, float]:
+    if device_type_id == 1:
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:75]: Please break this branch out into a separate function.

File: src/sensorpush_ble/parser.py [hunk 2]
```diff
@@ -40,18 +41,56 @@
...
+            + ((mfg_data[2] & 255) << 4)
+            + ((mfg_data[3] & 3) << 12)
+        )
+
+        return {
+            SensorLibrary.TEMPERATURE__CELSIUS: temperature_celsius,
+            SensorLibrary.HUMIDITY__PERCENTAGE: relative_humidity,
+        }
+
     """Decode values."""
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:94]: Please fix the doc string

File: src/sensorpush_ble/parser.py [hunk 3]
```diff
@@ -93,19 +132,27 @@ def _start_update(self, service_info: BluetoothServiceInfo) -> None:
...
-        for match_name in LOCAL_NAMES:
-            if match_name in local_name:
-                device_type = match_name
+        is_ht1 = False
+
         if (
-            not device_type
-            and "ef090000-11d6-42ba-93b8-9dd7ec090ab0" in service_info.service_uuids
+            local_name == "s"
+            and "ef090000-11d6-42ba-93b8-9dd7ec090aa9" in service_info.service_uuids
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:139]: Let's make the uuids constants so they aren't magic numbers that we won't remember later

**Comment by ashleigh-hopkins** [comment]
Hey @bdraco I believe all comments have been addressed and I've added a couple more tests to improve the code coverage

File: src/sensorpush_ble/parser.py [hunk 4]
```diff
@@ -16,21 +16,25 @@
...
 SENSORPUSH_MANUFACTURER_DATA_LEN = {
     3: "HT.w",
     5: "HTP.xw",
 }
 
-LOCAL_NAMES = ["HTP.xw", "HT.w"]
+LOCAL_NAMES = ["s", "HT1", "HTP.xw", "HT.w"]
+
+SENSORPUSH_UUID = "ef090000-11d6-42ba-93b8-9dd7ec090aa9"
+SENSORPUSH_UUID_V2 = "ef090000-11d6-42ba-93b8-9dd7ec090ab0"
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:30]: Please name the constants to show intent

https://en.wikipedia.org/wiki/Magic_number_(programming)

> Names chosen to be meaningful in the context of the program can result in code that is more easily understood by a maintainer who is not the original author (or even by the original author after a period of time).[[5]](https://en.wikipedia.org/wiki/Magic_number_(programming)#cite_note-Paul_2002_SYMBOLS-5) An example of an uninformatively named constant is int SIXTEEN = 16, while int NUMBER_OF_BITS = 16 is more descriptive.

File: src/sensorpush_ble/parser.py [hunk 5]
```diff
@@ -118,16 +173,19 @@ def _start_update(self, service_info: BluetoothServiceInfo) -> None:
...
+        if data := _find_latest_data(changed_manufacturer_data, is_ht1):
+            device_type_id = 1 if is_ht1 else 64 + (data[0] >> 2)
             if known_device_type := SENSORPUSH_DEVICE_TYPES.get(device_type_id):
                 device_type = known_device_type
             result.update(decode_values(data, device_type_id))
 
-        if service_info.name.startswith("SensorPush "):
-            self.set_device_name(service_info.name[11:])
+        if is_ht1:
+            self.set_device_name(f"HT1 {service_info.address[12:].replace(':', '')}")
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:183]: Please use `short_address` here 
https://github.com/Bluetooth-Devices/bluetooth-data-tools/blob/e260e5b51d678131861320a49a06595c5742ed07/src/bluetooth_data_tools/__init__.py#L20

File: src/sensorpush_ble/parser.py [hunk 6]
```diff
@@ -93,19 +145,22 @@ def _start_update(self, service_info: BluetoothServiceInfo) -> None:
...
+            for match_name in LOCAL_NAMES:
+                if match_name in local_name:
+                    device_type = match_name
+            if not device_type and (SENSORPUSH_UUID_V2 in service_info.service_uuids):
+                first_manufacturer_data_value_len = len(
+                    next(iter(manufacturer_data.values()))
+                )
+                device_type = SENSORPUSH_MANUFACTURER_DATA_LEN.get(
+                    first_manufacturer_data_value_len
+                )
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:163]: This could flow a bit better with returns if you made this a function `determine_device_type`

**Comment by bdraco** [comment]
Looks good. I added a few more review comments above

File: src/sensorpush_ble/parser.py (hunk 4 above)
Comment (ashleigh-hopkins) [reply src/sensorpush_ble/parser.py:30]: Can you let me know what you would like these values to be? I am just trying to help here, and I'm not aware of all the contribution requirements for your repo

File: src/sensorpush_ble/parser.py (hunk 5 above)
Comment (ashleigh-hopkins) [reply src/sensorpush_ble/parser.py:183]: Means bumping the project up to requring python 3.10, but I've done it

File: pyproject.toml [hunk 7]
```diff
@@ -23,14 +23,15 @@ packages = [
...
-python = "^3.9"
+python = "^3.10"
 
 # Documentation Dependencies
 Sphinx = {version = "^5.0", optional = true}
 sphinx-rtd-theme = {version = "^1.0", optional = true}
 myst-parser = {version = "^0.18", optional = true}
 bluetooth-sensor-state-data = ">=1.6.1"
 sensor-state-data = ">=2.0.2"
+bluetooth-data-tools = "^1.19.0"
```
Comment (bdraco) [inline pyproject.toml:34]: Please use >= here

File: src/sensorpush_ble/parser.py [hunk 8]
```diff
@@ -82,6 +135,31 @@ def decode_values(
     return values
 
 
+def determine_device_type(
+    service_info: BluetoothServiceInfo,
+    manufacturer_data: dict[int, bytes]
+) -> str:
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:141]: str | None

File: src/sensorpush_ble/parser.py [hunk 9]
```diff
@@ -82,6 +135,31 @@ def decode_values(
...
+
+    if not device_type and (SENSORPUSH_SERVICE_UUID_V2 in service_info.service_uuids):
+        first_manufacturer_data_value_len = len(
+            next(iter(manufacturer_data.values()))
+        )
+        return SENSORPUSH_MANUFACTURER_DATA_LEN.get(
+            first_manufacturer_data_value_len
+        )
+
+    return ""
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:161]: return None here

File: src/sensorpush_ble/parser.py [hunk 10]
```diff
@@ -82,6 +135,31 @@ def decode_values(
     return values
 
 
+def determine_device_type(
+    service_info: BluetoothServiceInfo,
+    manufacturer_data: dict[int, bytes]
+) -> str:
+    """Determine the device type based on the name and UUID"""
+    device_type = None
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:143]: Looks unused now

File: src/sensorpush_ble/parser.py [hunk 11]
```diff
@@ -82,6 +135,31 @@ def decode_values(
...
+    local_name = service_info.name
+
+    if local_name == "s" and SENSORPUSH_SERVICE_UUID_HT1 in service_info.service_uuids:
+        return "HT1"
+
+    for match_name in LOCAL_NAMES:
+        if match_name in local_name:
+            return match_name
+
+    if not device_type and (SENSORPUSH_SERVICE_UUID_V2 in service_info.service_uuids):
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:153]: ()s look unneeded

**Comment by bdraco** [comment]
I ordered an HT1 to test with.  Should be here sometime after the new year

File: src/sensorpush_ble/parser.py [hunk 12]
```diff
@@ -9,28 +9,33 @@
...
-SENSORPUSH_DEVICE_TYPES = {64: "HTP.xw", 65: "HT.w"}
+SENSORPUSH_DEVICE_TYPES = {1: "HT1", 64: "HTP.xw", 65: "HT.w"}
 
 SENSORPUSH_MANUFACTURER_DATA_LEN = {
     3: "HT.w",
     5: "HTP.xw",
 }
 
-LOCAL_NAMES = ["HTP.xw", "HT.w"]
+LOCAL_NAMES = ["s", "HT1", "HTP.xw", "HT.w"]
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:27]: It looks like this change isn't needed?

**Comment by bdraco** [comment]
The ci is really out of date for this library. I'm going to need to fix that before this can be released

File: src/sensorpush_ble/parser.py (hunk 4 above)
Comment (bdraco) [reply src/sensorpush_ble/parser.py]: new names LGTM

File: src/sensorpush_ble/parser.py [hunk 13]
```diff
@@ -40,19 +45,67 @@
...
+
+
+def temperature_celsius_from_raw_temperature(num: int) -> float:
+    return round((-46.85) + (175.72 * (num / (pow(2.0, 14.0)))), 2)
+
+
+def decode_ht1_values(mfg_data: bytes) -> dict[BaseSensorDescription, float]:
+    """Decode values for HT1."""
+    if mfg_data is None or len(mfg_data) == 0:
+        return {}
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:79]:
```suggestion
    if not mfg_data:
        return {}
```

Comment (bdraco) [reply src/sensorpush_ble/parser.py:79]: Can't be `None` here

File: src/sensorpush_ble/parser.py [hunk 14]
```diff
@@ -40,19 +45,67 @@
...
+
+
+def temperature_celsius_from_raw_temperature(num: int) -> float:
+    return round((-46.85) + (175.72 * (num / (pow(2.0, 14.0)))), 2)
+
+
+def decode_ht1_values(mfg_data: bytes) -> dict[BaseSensorDescription, float]:
+    """Decode values for HT1."""
+    if not mfg_data:
+        return {}
```
Comment (bdraco) [inline src/sensorpush_ble/parser.py:79]:
```suggestion
    if not len(mfg_data) < 4:
        return {}
```

**Review by bdraco — APPROVED** [review]

**Comment by bdraco** [comment]
Thanks @ashleigh-hopkins

**Comment by bdraco** [comment]
https://pypi.org/project/sensorpush-ble/ 1.6.1 published

**Pull request merged**